src/backend/distributed/executor/adaptive_executor.c

```diff
@@ -1171,16 +1171,15 @@ CleanUpSessions(DistributedExecution *execution)
 				elog(WARNING, "unexpected transaction state at the end of execution: %d",
 					 transactionState);
 			}
+
+			/* get ready for the next executions if we need use the same connection */
+			connection->waitFlags = WL_SOCKET_READABLE | WL_SOCKET_WRITEABLE;
 		}
 		else
 		{
 			elog(WARNING, "unexpected connection state at the end of execution: %d",
 				 connection->connectionState);
 		}
-
-		/* get ready for the next executions if we need use the same connection */
-		connection->waitFlags = WL_SOCKET_READABLE | WL_SOCKET_WRITEABLE;
-		execution->waitFlagsChanged = true;
```

Comment:
No need for this, we are done with the execution.

Reply:
That's not quite true for multi-row inserts. We don't need to set waitFlagsChanged because we will rebuild the WaitEventSet anyway. I do wonder where we set the wait flags to writeable before the next execution.

Reply:
Ah, it's the line you added :)